### PR TITLE
Refactor makeurl(), deprecate query taking string or list arguments, drop osc_urlencode()

### PR DESCRIPTION
**Pull request by dmach**

### CODE REVIEW

**Comment by dmach** [comment]
The failing `features/undelete.feature` behave test is expected.

**Review by dcermak — APPROVED** [review]
LGTM, I've added a few stylistic comments/nitpicks, but sadly I cannot estimate whether this will not break anything…

File: osc/core.py [hunk 1]
```diff
@@ -3606,39 +3607,72 @@ def pathjoin(a, *p):
...
+    :param path: List of URL path components.
+    :param query: Optional dictionary with URL query data.
+                  Values can be: ``str``, ``int``, ``bool``, ``[str]``, ``[int]``.
+                  Items with value equal to ``None`` will be skipped.
     """
-    query = query or []
-    _private.print_msg("makeurl:", baseurl, l, query, print_to="debug")
+    apiurl_scheme, apiurl_netloc, apiurl_path = urlsplit(apiurl)[0:3]
+
+    path = [i.decode("utf-8") if isinstance(i, bytes) else i for i in path]
```
[the remark this answers is not in the document]
Comment (dmach) [reply osc/core.py:3630]: Makes sense, I think it's no longer needed after removing `pathname2url(filename.encode(locale.getpreferredencoding(), "replace"))` from the code

File: osc/core.py [hunk 2]
```diff
@@ -3606,39 +3607,72 @@ def pathjoin(a, *p):
...
-    function. In case of a list not -- this is to be backwards compatible.
+    :param apiurl: URL to the API server.
+    :param path: List of URL path components.
+    :param query: Optional dictionary with URL query data.
+                  Values can be: ``str``, ``int``, ``bool``, ``[str]``, ``[int]``.
+                  Items with value equal to ``None`` will be skipped.
     """
-    query = query or []
-    _private.print_msg("makeurl:", baseurl, l, query, print_to="debug")
+    apiurl_scheme, apiurl_netloc, apiurl_path = urlsplit(apiurl)[0:3]
```
[the remark this answers is not in the document]
Comment (dmach) [reply osc/core.py:3628]: Let's keep it as it is.

**Pull request merged**